src/GafferML/DataToTensor.cpp

```diff
@@ -0,0 +1,189 @@
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferML/DataToTensor.h"
+
+#include "Gaffer/Context.h"
+#include "Gaffer/PlugAlgo.h"
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferML;
+
+GAFFER_NODE_DEFINE_TYPE( DataToTensor );
+
+size_t DataToTensor::g_firstPlugIndex = 0;
+const IECore::InternedString DataToTensor::g_dataPlugName( "data" );
+
+DataToTensor::DataToTensor( const std::string &name )
+	:	ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new IntPlug( "shapeMode", Plug::In, (int)ShapeMode::Automatic, (int)ShapeMode::Automatic, (int)ShapeMode::Custom ) );
+	addChild( new Int64VectorDataPlug( "shape" ) );
+	addChild( new TensorPlug( "tensor", Plug::Out ) );
+}
+
+DataToTensor::~DataToTensor()
+{
+}
+
+bool DataToTensor::canSetup( const Gaffer::ValuePlug *prototypeDataPlug )
+{
+	if( dataPlug() )
+	{
+		return false;
+	}
+
+	/// \todo Check type
+	return true;
+}
+
+void DataToTensor::setup( const Gaffer::ValuePlug *prototypeDataPlug )
+{
+	if( dataPlug() )
+	{
+		throw IECore::Exception( "DataToTensor already has a \"data\" plug." );
+	}
+
+	PlugPtr dataPlug = prototypeDataPlug->createCounterpart( g_dataPlugName, Plug::In );
+	dataPlug->setFlags( Plug::Serialisable, true );
+	addChild( dataPlug );
+}
+
+Gaffer::IntPlug *DataToTensor::shapeModePlug()
+{
+	return getChild<IntPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::IntPlug *DataToTensor::shapeModePlug() const
+{
+	return getChild<IntPlug>( g_firstPlugIndex );
+}
+
+Gaffer::Int64VectorDataPlug *DataToTensor::shapePlug()
+{
+	return getChild<Int64VectorDataPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::Int64VectorDataPlug *DataToTensor::shapePlug() const
+{
+	return getChild<Int64VectorDataPlug>( g_firstPlugIndex + 1 );
+}
+
+TensorPlug *DataToTensor::tensorPlug()
+{
+	return getChild<TensorPlug>( g_firstPlugIndex + 2 );
+}
+
+const TensorPlug *DataToTensor::tensorPlug() const
+{
+	return getChild<TensorPlug>( g_firstPlugIndex + 2 );
+}
+
+void DataToTensor::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	ComputeNode::affects( input, outputs );
+
+	if(
+		input == dataPlug() ||
+		input == shapeModePlug() ||
+		input == shapePlug()
+	)
+	{
+		outputs.push_back( tensorPlug() );
+	}
+}
+
+void DataToTensor::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	if( output == tensorPlug() )
+	{
+		ComputeNode::hash( output, context, h );
```

Comment:
I have done that in the past, but I'm adopting the current form as a better convention. Some hash functions want to assign directly to the hash (usually to denote a pass-through), in which case having called the base class outside the `if` means you've done redundant work that you're now overwriting. I also find it clearer when all the code for one case is inside a single `{}` block, rather than having to read across the conditional to see everything that is happening for `tensorPlug()`. It's easier to extend this new form with additional cases.

Reply:
Alright